container_images/pytest/main.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Execute pytest within a virtual environment."""
+
+import configparser
+import os
+import re
+import shutil
+import sys
+import tempfile
+import typing
+
+import toml
+import tox
+
+# Dependencies to install in addition to user's requested dependencies.
+_common_test_deps = ["pytest==6.0.1"]
+
+_per_interpreter_command = [
+  # Don't invoke the `pytest` binary directly, as it circumvents
+  # the virtual environment.
+  "python",
+  "-m",
+  "pytest",
+  # Write test result to stdout out, except if test passes.
+  "-ra",
+  # Ensure junit report uses xunit1, which is the syntax used
+  # by spyglass.
+  # https://github.com/GoogleCloudPlatform/testgrid/blob/master/metadata/junit/junit.go
+  "--override-ini=junit_family=xunit1",
+  # Within the junit report, add a namespace for the environment,
+  # to ensure that tests from different interpreters aren't
+  # considered duplicates.
+  "--junit-prefix={envname}",
+  # Write a new junit xml file for each interpreter. The filename pattern
+  # is defined in our Prow instance's configuration:
+  #  https://github.com/GoogleCloudPlatform/oss-test-infra/blob/cc1f0cf1ffecc0e3b75664b0d264abc6165276d1/prow/oss/config.yaml#L98
+  "--junit-xml={env:ARTIFACTS}/junit-{envname}.xml",
+]
+
+
+def setup_execution_root(pyproject_toml_file: str) -> str:
+  """Create execution directory and copy project's code to it.
+
+  Tox creates in-directory files, and its execution fails when the
+  source tree is mounted read-only.
```

Comment:
does prow make the source tree read only? i didn't know that...

Reply:
I couldn't find documentation either way regarding Prow, but when running locally, I found it useful to mount read-only to ensure `tox` wasn't leaving files on my local repo.